compiler/src/cobalt/Parser/Parser.hs

```diff
@@ -263,13 +263,14 @@ methodParser = do
           annotations <- many annotationParser
           modifiers <- modifiersParser
           rword "let"
-          name <- nameParser
+          name <- choice [Name <$> rword "this", nameParser]
           fields <- parens $ sepBy fieldParser $ symbol ","
           return (annotations, modifiers, name, fields)
-        symbol ":"
-        returnType <- typeRefParser
+        returnType <- optional $ do
+            symbol ":"
+            typeRefParser
         symbol "="
```

Comment:
This might allow parsing methods without return type that are not constructors. Is this expected? If not, we can have  `constructorStar` and `methodStart` in the `where` clause - the method one would be as it was and the constructor one would be hardwired for expecting `this` as the name of method and call them using `choice` where we call `methodStart` now.

Reply:
Yes this is expected. I need to write this down in the documentation. I feel type inference in some cases might be useful. Even simple cases such as `let x() = 10` would be a really nice feature.

Reply:
I am questioning something. Instead of adding a new type called `Init` I could define a new data constructor.
Which do you prefer? It's either we wrap it or set the type.
```
data Method
    = Constructor Method
    | Method
    { methodName :: Name
    , methodAnns :: [Annotation]
    , methodParams :: [Field]
    , methodModifiers :: [Modifier]
    , methodReturnType :: (Maybe Type)
    , methodBody :: Assignment
    }
    deriving (Show, Eq)
```

Reply:
We will see how much different the `genCode` will need to be for constructors compared to normal methods.

Reply:
If it's anything like generating code using ASM you define them very similar. 

This is a comparison of a constructor and the main method in Java ASM. 
```
        MethodVisitor constructor = classWriter.visitMethod(
                ACC_PUBLIC,
                "<init>",
                "()V",
                null,
                null);

        MethodVisitor main = classWriter.visitMethod(
                ACC_PUBLIC + ACC_STATIC,
                "main",
                "([Ljava/lang/String;)V",
                null,
                null
        );
```